File: toon/budget.py
"""Token budget allocation across tiers and entries.

Budget distribution:
  - "preserve" tier: full original size (no compression)
  - "high" tier: 60% of remaining budget (proportional to scores)
  - "medium" tier: 30% of remaining budget
  - "low" tier: 10% of remaining budget
  - "cut" tier: 0 tokens (excluded)

These ratios are engineering heuristics informed by LLMLingua's finding that
dynamic per-component allocation (EM 79.08) outperforms uniform (73.62).
Source: https://aclanthology.org/2024.acl-long.91/

Performance: O(n) for n entries. Budget calculation is pure arithmetic.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ._utils import estimate_tokens_obj


@dataclass
class BudgetAllocation:
    """Per-entry budget allocation result."""
    entry_budgets: list[int]
    tier_budgets: dict[str, int]
    total_budget: int
    reserve: int


class BudgetAllocator:
    """Allocate token budgets across entries based on tier and score.

    Two-phase allocation:
      1. Compute per-tier budgets from the tier ratios
      2. Distribute within each tier proportional to entry scores

    The "preserve" tier receives its exact original token count.
    The remaining budget is split 60/30/10 across high/medium/low.
    Each entry within a tier receives a share proportional to its score.
    """

    TIER_RATIOS: dict[str, float] = {
        "high": 0.60,
        "medium": 0.30,
        "low": 0.10,
    }
    OVERHEAD_RESERVE: float = 0.05  # 5% for structural markers

    @classmethod
    def allocate(
        cls,
        entries: list[dict],
        scores: list[float],
        tiers: list[str],
        total_budget: int,
    ) -> BudgetAllocation:
        """Allocate budget across entries.

        Args:
            entries: Entry metadata dicts with "content" key.
            scores: Normalized [0,1] scores per entry.
            tiers: Tier assignment per entry ("preserve"|"high"|"medium"|"low"|"cut").
            total_budget: Total available token budget.

        Returns:
            BudgetAllocation with per-entry budgets.
        """
        reserve = int(total_budget * cls.OVERHEAD_RESERVE)
        usable = total_budget - reserve

        # Calculate preserve tier consumption
        preserve_tokens = 0
        for i, tier in enumerate(tiers):
            if tier == "preserve":
                preserve_tokens += estimate_tokens_obj(entries[i]["content"])

        remaining = max(0, usable - preserve_tokens)

        # Allocate to tiers
        tier_budgets: dict[str, int] = {"preserve": preserve_tokens}
        for tier_name, ratio in cls.TIER_RATIOS.items():
            tier_budgets[tier_name] = int(remaining * ratio)
        tier_budgets["cut"] = 0

        # Distribute within each tier proportional to score
        entry_budgets: list[int] = [0] * len(entries)

        for tier_name in ["preserve", "high", "medium", "low", "cut"]:
            tier_indices = [i for i, t in enumerate(tiers) if t == tier_name]
            if not tier_indices:
                continue

            if tier_name == "preserve":
                for i in tier_indices:
                    entry_budgets[i] = estimate_tokens_obj(entries[i]["content"])
                continue

            if tier_name == "cut":
                continue

            tier_budget = tier_budgets[tier_name]
            tier_scores = [max(scores[i], 1e-10) for i in tier_indices]
            score_sum = sum(tier_scores)

            for idx, i in enumerate(tier_indices):
                share = (
                    tier_scores[idx] / score_sum
                    if score_sum > 0
                    else 1.0 / len(tier_indices)
                )
                entry_budgets[i] = max(10, int(tier_budget * share))

        return BudgetAllocation(
            entry_budgets=entry_budgets,
            tier_budgets=tier_budgets,
            total_budget=total_budget,
            reserve=reserve,
        )

File: src/budget.ts
import { estimateTokensObj } from './utils.js';
import { EntryMeta } from './types.js';

// ---------------------------------------------------------------------------
// BudgetAllocation
// ---------------------------------------------------------------------------

/** Per-entry budget allocation result. */
export interface BudgetAllocation {
  entry_budgets: number[];
  tier_budgets: Record<string, number>;
  total_budget: number;
  reserve: number;
}

// ---------------------------------------------------------------------------
// BudgetAllocator
// ---------------------------------------------------------------------------

/**
 * Allocate token budgets across entries based on tier and score.
 *
 * Two-phase allocation:
 *   1. Compute per-tier budgets from the tier ratios
 *   2. Distribute within each tier proportional to entry scores
 *
 * The "preserve" tier receives its exact original token count.
 * The remaining budget is split 60/30/10 across high/medium/low.
 * Each entry within a tier receives a share proportional to its score.
 */
export class BudgetAllocator {
  static readonly TIER_RATIOS: Record<string, number> = {
    high: 0.60,
    medium: 0.30,
    low: 0.10,
  };

  static readonly OVERHEAD_RESERVE: number = 0.05; // 5% for structural markers

  /**
   * Allocate budget across entries.
   *
   * @param entries Entry metadata dicts with "content" key.
   * @param scores Normalized [0,1] scores per entry.
   * @param tiers Tier assignment per entry ("preserve"|"high"|"medium"|"low"|"cut").
   * @param total_budget Total available token budget.
   * @returns BudgetAllocation with per-entry budgets.
   */
  static allocate(
    entries: EntryMeta[],
    scores: number[],
    tiers: string[],
    total_budget: number
  ): BudgetAllocation {
    const reserve = Math.trunc(total_budget * BudgetAllocator.OVERHEAD_RESERVE);
    const usable = total_budget - reserve;

    // Calculate preserve tier consumption
    let preserve_tokens = 0;
    for (const [i, tier] of tiers.entries()) {
      if (tier === 'preserve') {
        const entry = entries[i];
        if (entry === undefined) {
          throw new Error(
            `invariant: entries[${i}] missing for tier index in preserve loop`
          );
        }
        preserve_tokens += estimateTokensObj(entry.content);
      }
    }

    const remaining = Math.max(0, usable - preserve_tokens);

    // Allocate to tiers
    const tier_budgets: Record<string, number> = {
      preserve: preserve_tokens,
    };
    for (const tier_name of Object.keys(BudgetAllocator.TIER_RATIOS)) {
      const ratio = BudgetAllocator.TIER_RATIOS[tier_name];
      if (ratio === undefined) {
        throw new Error(
          `invariant: TIER_RATIOS missing entry for tier '${tier_name}'`
        );
      }
      tier_budgets[tier_name] = Math.trunc(remaining * ratio);
    }
    tier_budgets['cut'] = 0;

    // Distribute within each tier proportional to score
    const entry_budgets: number[] = new Array<number>(entries.length).fill(0);

    for (const tier_name of ['preserve', 'high', 'medium', 'low', 'cut']) {
      const tier_indices: number[] = [];
      for (const [i, tier] of tiers.entries()) {
        if (tier === tier_name) {
          tier_indices.push(i);
        }
      }
      if (tier_indices.length === 0) {
        continue;
      }

      if (tier_name === 'preserve') {
        for (const i of tier_indices) {
          const entry = entries[i];
          if (entry === undefined) {
            throw new Error(
              `invariant: entries[${i}] missing for preserve tier index`
            );
          }
          entry_budgets[i] = estimateTokensObj(entry.content);
        }
        continue;
      }

      if (tier_name === 'cut') {
        continue;
      }

      const tier_budget = tier_budgets[tier_name];
      if (tier_budget === undefined) {
        throw new Error(
          `invariant: no budget for tier '${tier_name}'`
        );
      }
      const tier_scores: number[] = tier_indices.map((i) => {
        const s = scores[i];
        if (s === undefined) {
          throw new Error(
            `invariant: scores[${i}] missing for tier index in '${tier_name}'`
          );
        }
        return Math.max(s, 1e-10);
      });
      const score_sum = tier_scores.reduce((a, b) => a + b, 0);

      for (const [idx, i] of tier_indices.entries()) {
        const tier_score = tier_scores[idx];
        if (tier_score === undefined) {
          throw new Error(
            `invariant: tier_scores[${idx}] missing in tier '${tier_name}'`
          );
        }
        const share =
          score_sum > 0
            ? tier_score / score_sum
            : 1.0 / tier_indices.length;
        entry_budgets[i] = Math.max(10, Math.trunc(tier_budget * share));
      }
    }

    return {
      entry_budgets,
      tier_budgets,
      total_budget,
      reserve,
    };
  }
}
